File: solutions/go/src/playground/algorithm/n_queens.go
package algorithm

func solveNQueens(n int) [][]string {
	result := make([][]string, 0)

	// 初始化棋盘
	board := make([][]bool, n)
	for i := 0; i < n; i++ {
		board[i] = make([]bool, n)
	}

	// 递归调用
	solveNQueensR(board, n, n, 0, &result)

	return result
}

func solveNQueensR(board [][]bool, width, height, row int, result *[][]string) {
	if row == height {
		solution := make([]string, height)
		for i, r := range board {
			s := ""
			for _, c := range r {
				if c {
					s += "Q"
				} else {
					s += "."
				}
			}
			solution[i] = s
		}
		*result = append(*result, solution)
		return
	}

	for c := 0; c < width; c++ {
		board[row][c] = true
		if !hasConflict(board, width, height, row, c) {
			solveNQueensR(board, width, height, row+1, result)
		}
		board[row][c] = false
	}
}

func hasConflict(board [][]bool, width, height, row, column int) bool {
	for i := 0; i < row; i++ {
		if board[i][column] {
			return true
		}
	}

	for i := row + 1; i < width; i++ {
		if board[i][column] {
			return true
		}
	}

	for i := 0; i < column; i++ {
		if board[row][i] {
			return true
		}
	}

	for i := column + 1; i < height; i++ {
		if board[row][i] {
			return true
		}
	}

	for i, j := row-1, column-1; i >= 0 && j >= 0; i, j = i-1, j-1 {
		if board[i][j] {
			return true
		}
	}

	for i, j := row+1, column+1; i < width && j < height; i, j = i+1, j+1 {
		if board[i][j] {
			return true
		}
	}

	for i, j := row-1, column+1; i >= 0 && j < height; i, j = i-1, j+1 {
		if board[i][j] {
			return true
		}
	}

	for i, j := row+1, column-1; i < width && j >= 0; i, j = i+1, j-1 {
		if board[i][j] {
			return true
		}
	}

	return false
}

File: solutions/java/src/main/java/playground/algorithm/NQueens.java
package playground.algorithm;

import java.util.ArrayList;
import java.util.List;

public class NQueens {

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        solveNQueens(0, n, new ArrayList<>(), result);
        return result;
    }

    private void solveNQueens(int row, int size, List<String> board, List<List<String>> boards) {
        if (row == size) {
            boards.add(new ArrayList<>(board));
            return;
        }
        for (int column = 0; column < size; column++) {
            if (hasConflict(row, column, size, board)) {
                continue;
            }
            board.add(getRow(column, size));
            solveNQueens(row + 1, size, board, boards);
            board.remove(board.size() - 1);
        }
    }

    private boolean hasConflict(int row, int column, int size, List<String> board) {
        for (int r = 0; r < row; r++) {
            if (board.get(r).charAt(column) == 'Q') {
                return true;
            }
        }

        for (int r = row - 1, c = column - 1; r >= 0 && c >= 0; r--, c--) {
            if (board.get(r).charAt(c) == 'Q') {
                return true;
            }
        }

        for (int r = row - 1, c = column + 1; r >= 0 && c < size; r--, c++) {
            if (board.get(r).charAt(c) == 'Q') {
                return true;
            }
        }

        return false;
    }

    private String getRow(int column, int size) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < column; i++) {
            builder.append(".");
        }
        builder.append("Q");
        for (int i = column + 1; i < size; i++) {
            builder.append(".");
        }
        return builder.toString();
    }
}
